packages/server/src/utils/ldapLogin.ts

```diff
@@ -0,0 +1,77 @@
+import { Client } from 'ldapts';
+import { User } from '../services/database/types';
+import Logger from '@joplin/utils/Logger';
+import { LdapConfig } from './types';
+import { ErrorForbidden } from './errors';
+
+const logger = Logger.create('LDAP');
+
+export default async function ldapLogin(email: string, password: string, user: User, config: LdapConfig): Promise<User> {
+
+	const enabled = config.enabled;
+	const userCreation = config.userCreation;
+	const host = config.host;
+	const mailAttribute = config.mailAttribute;
+	const fullNameAttribute = config.fullNameAttribute;
+	const baseDN = config.baseDN;
+	const bindDN = config.bindDN;
+	const bindPW = config.bindPW;
+
+	logger.info(`Starting authentication with Server ${host}`);
+
+	if (password === '') {
+		throw new ErrorForbidden('no password entered');
+	}
+
+	if (enabled) {
+		let searchResults;
+		const client = new Client({
+			url: host,
+			timeout: 2000,
+			connectTimeout: 1000,
+		});
+
+		if (bindDN.length !== 0) {
+			try {
+				await client.bind(bindDN, bindPW);
+			} catch (error) {
+				error.message = `Could not bind to the ldap server ${host}: ${error.message}`;
+				throw error;
+			}
+		}
+
+		try {
+			searchResults = await client.search(baseDN, {
+				filter: `(${mailAttribute}=${email})`,
+				attributes: ['dn', fullNameAttribute],
+			});
+		} catch (error) {
+			error.message = `Could not search the ldap server ${host}: ${error.message}`;
+			throw error;
+		}
```

Comment:
What's the logic here - if the entries cannot be found does it return an empty array, or does it throw an exception?

If it throws an exception, I think the flow should be the same as below - return `null` if the email could not be found, or rethrow if it's a server-side issue.

Reply:
It returns an empty array. And this leads to an error when trying to bind in the next step.
I can add a check and return null if the user was not found.

Reply:
Yes returning null at this point, when getting an empty array, would make sense.